### PR TITLE
Added a cookbook section about event subscribers

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | all |
| Fixed tickets | - |

This PR finishes the work made by @beni0888 in #4538.

### CODE REVIEW

File: cookbook/service_container/event_listener.rst [hunk 1]
```diff
@@ -1,27 +1,36 @@
 .. index::
    single: Events; Create listener
+   single: Create subscriber
 
-How to Create an Event Listener
-===============================
+How to Create Event Listeners and Subscribers
+=============================================
 
 Symfony has various events and hooks that can be used to trigger custom
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: [...] to perform custom [...]

File: cookbook/service_container/event_listener.rst [hunk 2]
```diff
@@ -1,27 +1,36 @@
...
-behavior in your application. Those events are thrown by the HttpKernel
-component and can be viewed in the :class:`Symfony\\Component\\HttpKernel\\KernelEvents` class.
+actions in your application. Those events are thrown by the HttpKernel
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: [...] are triggered by [...]

File: cookbook/service_container/event_listener.rst [hunk 3]
```diff
@@ -1,27 +1,36 @@
...
-you to modify how exceptions are shown by your application. The ``KernelEvents::EXCEPTION``
-event is just one of the core kernel events::
+To hook into an event and execute your own custom logic, you have to create
+a service that listens to that event. As explained in this article, you can do
+that in two different ways: creating an event listener or an event subscriber.
 
-    // src/AppBundle/EventListener/AcmeExceptionListener.php
-    namespace AppBundle\EventListener;
+The examples of this article only use the ``KernelEvents::EXCEPTION`` event for
+consistency purposes. In your own application you can use any event and even mix
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: [...] application, you [...]

File: cookbook/service_container/event_listener.rst [hunk 4]
```diff
@@ -71,53 +77,153 @@ using a special "tag":
 
     .. code-block:: yaml
 
-        # app/config/services.yml
+        # app/config/config.yml
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: What is the reason for this change?

File: cookbook/service_container/event_listener.rst [hunk 5]
```diff
@@ -71,53 +77,153 @@ using a special "tag":
...
+            ->register('kernel.listener.your_listener_name', 'AppBundle\Listener\ExceptionListener')
+            ->addTag('kernel.event_listener', array('event' => 'kernel.exception'))
         ;
 
 .. note::
 
-    There is an additional tag option ``priority`` that is optional and defaults
-    to 0. The listeners will be executed in the order of their priority (highest to lowest).
-    This is useful when you need to guarantee that one listener is executed before another.
+    There is an optional tag option called ``method`` which defines which method
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: I would say "tag attribute".

File: cookbook/service_container/event_listener.rst [hunk 6]
```diff
@@ -71,53 +77,153 @@ using a special "tag":
...
+    to execute when the event is triggered. By default the name of the method is
+    ``on`` + "camel-cased event name". If the event is ``kernel.exception`` the
+    method executed by default is ``onKernelException()``.
+
+    The other optional tag option is called  ``priority`` and it defaults to ``0``.
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: [...] which defaults [...]

File: cookbook/service_container/event_listener.rst [hunk 7]
```diff
@@ -71,53 +77,153 @@ using a special "tag":
...
+    This value ranges from ``-255`` to ``255`` and it controls the order in which
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: Are there really any limits besides the range of the `int` type? IIRC, these are only the minimum and maximum priorities used in the core.

File: cookbook/service_container/event_listener.rst [hunk 8]
```diff
@@ -71,53 +77,153 @@ using a special "tag":
...
+    listeners are executed (the highest the priority, the earlier a listener is
+    executed). This is useful when you need to guarantee that one listener is
+    executed before another.
+
+Creating an Event Subscriber
+----------------------------
+
+Another way to listen to events is via an **event subscriber**, which is a class
+that can define one or more methods that listen to one or various events. The
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: The real difference between a pure listener and a subscriber is that the subscriber knows to which events it actually listens to.

File: cookbook/service_container/event_listener.rst [hunk 9]
```diff
@@ -71,53 +77,153 @@ using a special "tag":
...
+the method is called). To learn more about event subscribers, read :doc:`/components/event_dispatcher/introduction`.
+The following example shows an event subscriber that defines several methods which
+listen to the same ``kernel.exception`` event::
+
+    // src/AppBundle/Subscriber/ExceptionSubscriber.php
+    namespace AppBundle\Subscriber;
+
+    use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+    use Symfony\Component\HttpKernel\Event\GetResponseForExceptionEvent;
+    use Symfony\Component\HttpFoundation\Response;
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: should be moved before the HttpKernel imports

File: cookbook/service_container/event_listener.rst [hunk 10]
```diff
@@ -71,53 +77,153 @@ using a special "tag":
...
+                    - { name: kernel.event_subscriber }
+
+    .. code-block:: xml
+
+        <!-- app/config/config.xml -->
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services">
+
+            <services>
+                <service id="acme_exception_subscriber"
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: wrong service id

File: cookbook/service_container/event_listener.rst [hunk 11]
```diff
@@ -71,53 +77,153 @@ using a special "tag":
...
+                </service>
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        // app/config/config.php
+        $container
+            ->register(
+                'acme_exception_subscriber',
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: same here

File: cookbook/service_container/event_listener.rst [hunk 12]
```diff
@@ -51,18 +60,15 @@ event is just one of the core kernel events::
         }
     }
 
+.. versionadded:: 2.4
+    Support for HTTP status code constants was introduced in Symfony 2.4.
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: Needs to be removed and added when this was merged up to the `2.6` branch.

File: cookbook/service_container/event_listener.rst [hunk 13]
```diff
@@ -71,53 +77,153 @@ using a special "tag":
...
+            )
+            ->addTag('kernel.event_subscriber')
+        ;
 
 Request Events, Checking Types
 ------------------------------
 
+.. versionadded:: 2.4
+    The ``isMasterRequest()`` method was introduced in Symfony 2.4.
+    Prior, the ``getRequestType()`` method must be used.
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: Needs to be removed and added when this was merged up to the `2.6` branch.

File: cookbook/service_container/event_listener.rst [hunk 14]
```diff
@@ -131,3 +237,5 @@ done as follow::
     Two types of request are available in the :class:`Symfony\\Component\\HttpKernel\\HttpKernelInterface`
     interface: ``HttpKernelInterface::MASTER_REQUEST`` and
     ``HttpKernelInterface::SUB_REQUEST``.
+
+.. _`The EventDispatcher component`: http://symfony.com/doc/current/components/event_dispatcher/index.html
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: should be removed

File: cookbook/service_container/event_listener.rst (hunk 4 above)
Comment (javiereguiluz) [inline cookbook/service_container/event_listener.rst]: This is an error. Reverted. Thanks.

File: cookbook/service_container/event_listener.rst [hunk 15]
```diff
@@ -74,30 +77,126 @@ using a special "tag":
...
+    }
+
+Now, you just need to register the class as a service and notify Symfony that it
+is an event subscriber:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # app/config/config.yml
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: Should be `services.yml` here too (same below for the other formats).

Comment (javiereguiluz) [inline cookbook/service_container/event_listener.rst]: Fixed. Thanks.

File: cookbook/service_container/event_listener.rst (hunk 8 above)
Comment (javiereguiluz) [inline cookbook/service_container/event_listener.rst]: You are right. I've reworded the introduction of this section. Thanks.

File: cookbook/service_container/event_listener.rst [hunk 16]
```diff
@@ -74,30 +77,132 @@ using a special "tag":
...
-    There is an additional tag option ``priority`` that is optional and defaults
-    to 0. The listeners will be executed in the order of their priority (highest to lowest).
-    This is useful when you need to guarantee that one listener is executed before another.
+    There is an optional tag attribute called ``method`` which defines which method
+    to execute when the event is triggered. By default the name of the method is
+    ``on`` + "camel-cased event name". If the event is ``kernel.exception`` the
+    method executed by default is ``onKernelException()``.
+
+    The other optional tag attribute is called  ``priority`` which defaults to ``0``.
+    This value ranges from ``-255`` to ``255`` and it controls the order in which
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: The possible values are not limited to this range. This is just the interval from which priorities inside the core are chosen.

File: cookbook/service_container/event_listener.rst [hunk 17]
```diff
@@ -74,30 +77,132 @@ using a special "tag":
...
+----------------------------
+
+Another way to listen to events is via an **event subscriber**, which is a class
+that defines one or more methods that listen to one or various events. The main
+difference with the event listeners is that subscribers always know which events
+they are listening to.
+
+In a given subscriber, different methods can listen to the same event. The order
+in which methods are executed is defined by the ``priority`` parameter of each
+method (the higher the priority, the earlier the method is called). To learn more
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: no need for this comma

**Comment by javiereguiluz** [comment]
All issues have been fixed. Can we label this as `finished`? Thanks!

File: cookbook/service_container/event_listener.rst [hunk 18]
```diff
@@ -1,27 +1,36 @@
...
+How to Create Event Listeners and Subscribers
+=============================================
 
-Symfony has various events and hooks that can be used to trigger custom
-behavior in your application. Those events are thrown by the HttpKernel
-component and can be viewed in the :class:`Symfony\\Component\\HttpKernel\\KernelEvents` class.
+Symfony has various events and hooks that can be used to perform custom
+actions in your application. Those events are triggered by the HttpKernel
+component and they are defined in the :class:`Symfony\\Component\\HttpKernel\\KernelEvents`
+class.
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: Imo this paragraph contradicts a bit your other PR where you try to clarify that user's don't have to provider their own event dispatcher for custom events.

Comment (javiereguiluz) [inline cookbook/service_container/event_listener.rst]: I'm stuck at this. Can someone please provide me an alternative to this paragraph? Thanks.

Comment (beni0888) [inline cookbook/service_container/event_listener.rst]: IMHO this paragraph is ok. I don't see any problem here.

**Comment by javiereguiluz** [comment]
I've updated this PR to add a note comparing listeners and subscribers, as asked in #3987.

**Comment by javiereguiluz** [comment]
Please, tell me what else I can do to get this PR labelled as `finished`. Thanks.

File: cookbook/service_container/event_listener.rst [hunk 19]
```diff
@@ -74,30 +77,133 @@ using a special "tag":
...
+
+            <services>
+                <service id="app.exception_subscriber"
+                    class="AppBundle\Subscriber\ExceptionSubscriber">
+
+                    <tag name="kernel.event_subscriber"/>
+
+                </service>
+            </services>
+        </container>
```
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]:
```xml
<!-- app/config/services.xml -->
<?xml version="1.0" encoding="UTF-8" ?>
<container xmlns="http://symfony.com/schema/dic/services"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">

    <services>
        <service id="app.exception_subscriber"
            class="AppBundle\Subscriber\ExceptionSubscriber">

            <tag name="kernel.event_subscriber"/>
        </service>
    </services>
</container>
```

File: cookbook/service_container/event_listener.rst (hunk 18 above)
Comment (xabbuh) [inline cookbook/service_container/event_listener.rst]: Probably, adding a short note after this paragraph that you can dispatch your own events using the core event dispatcher and that there are lots of more events provided by third-party libraries, could be sufficient (we can link to related articles).

File: cookbook/service_container/event_listener.rst [hunk 20]
```diff
@@ -74,30 +77,133 @@ using a special "tag":
...
+    There is an optional tag attribute called ``method`` which defines which method
+    to execute when the event is triggered. By default the name of the method is
+    ``on`` + "camel-cased event name". If the event is ``kernel.exception`` the
+    method executed by default is ``onKernelException()``.
+
+    The other optional tag attribute is called  ``priority``, which defaults to
+    ``0`` and it controls the order in which listeners are executed (the highest
+    the priority, the earlier a listener is executed). This is useful when you
+    need to guarantee that one listener is executed before another. The priorities
+    of the internal Symfony events range from ``-255`` to ``255`` but your own
```
Comment (stof) [inline cookbook/service_container/event_listener.rst]: events don't have a priority. Listeners do. And we have listeners with a priority higher than 255 in the core, so the message is wrong

File: cookbook/service_container/event_listener.rst [hunk 21]
```diff
@@ -74,30 +77,133 @@ using a special "tag":
...
+    events can use any positive or negative integer.
```
Comment (stof) [inline cookbook/service_container/event_listener.rst]: events don't use a priority. Listeners do

**Comment by OskarStark** [comment]
the PR title is weird.

this would be better IMO:
```Added a cookbook SECTION about event subscribers```

**Comment by javiereguiluz** [comment]
I've reworded the introduction and also made the required changes. This is ready for the final review. Thanks!

**Comment by webda2l** [comment]
In http://symfony.com/doc/current/cookbook/bundles/best_practices.html, directory for events listeners is ````EventListener/````, and ````Listener/```` in this PR.
Will be nice to rule on it.

**Comment by xabbuh** [comment]
Indeed, can you make that last change, Javier? Then, this is ready to be merged.

**Comment by javiereguiluz** [comment]
I've made some final changes in b5a82ca.

**Comment by xabbuh** [comment]
:+1: Thanks Javier. That's a really nice work and I think we are finally finished here. :)

**Comment by weaverryan** [comment]
Thanks Javier - and sorry for the extreme delay - I personally am catching up after some time away :)

**Comment by xabbuh** [comment]
@javiereguiluz During merging this up the branches, something seems to have been go wrong. I fixed the build in #5792. Would you mind checking if everything you added here is still present and correct?

**Comment by javiereguiluz** [comment]
I've checked and, at least in 2.3 branch, the article contains all the proposed changes, so nothing got lost.

**Comment by xabbuh** [comment]
@javiereguiluz It was broken on the `2.7` branch.
